---
title: "Modeling transition/transversion bias of nucleotide substitution over time"
description: "Case study 4"
author:
  - name: Stefan Schmutz 
    url: https://example.com/norajones
    affiliation: ZHAW
date: "`r Sys.Date()`"
output:
  distill::distill_article:
    toc: true
    toc_depth: 2
knit: pagedown::chrome_print

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(tidyverse)
library(seqinr)
library(knitr)
library(kableExtra)
```

```{r read-fasta, echo=FALSE}
mt_cyb <-
  read.fasta(file = "mt-cyb-human-mouse_cDNAalignment.fasta")

mt_cyb_df <-
  bind_cols("human" = mt_cyb$HOMOSAPIENSCYTB, "mouse" = mt_cyb$MUSMUSCULUSCYTB)
```


# Estimating nucleotide frequencies

> Estimate nucleotide frequencies from the pairwise alignment of human and mouse cytochrome b gene as given in the file “mt-cyb-human- mouse_cDNAalignment.fasta”. Use these values to parameterize the model.

One way to estimate the nucleotide frequencies is to count them.  
Since we're working with a pairwise alignment without indels, the total length
of both sequences is the same (`r length(mt_cyb$HOMOSAPIENSCYTB)` nt). 
The detailed composition is listed in Table \@ref(tab:nucleotide-frequencies).


```{r nucleotide-frequencies, echo=FALSE}

nt_T_human <- sum(mt_cyb$HOMOSAPIENSCYTB == "t")
nt_C_human <- sum(mt_cyb$HOMOSAPIENSCYTB == "c")
nt_A_human <- sum(mt_cyb$HOMOSAPIENSCYTB == "a")
nt_G_human <- sum(mt_cyb$HOMOSAPIENSCYTB == "g")

nt_T_mouse <- sum(mt_cyb$MUSMUSCULUSCYTB == "t")
nt_C_mouse <- sum(mt_cyb$MUSMUSCULUSCYTB == "c")
nt_A_mouse <- sum(mt_cyb$MUSMUSCULUSCYTB == "a")
nt_G_mouse <- sum(mt_cyb$MUSMUSCULUSCYTB == "g")

nt_frequencies <-
  tribble(
    ~nucleotide, ~organism, ~count,
    "T", "human", nt_T_human,
    "T", "mouse", nt_T_mouse,
    "C", "human", nt_C_human,
    "C", "mouse", nt_C_mouse,
    "A", "human", nt_A_human,
    "A", "mouse", nt_A_mouse,
    "G", "human", nt_G_human,
    "G", "mouse", nt_G_mouse
  )

nt_frequencies %>%
  pivot_wider(names_from = "organism", values_from = "count") %>%
  kable(caption = "Nucleotide frequencies") %>%
  kable_styling(full_width = F)

```

# Estimating transition transversion rate ratio

> Propose a simple way of estimating transition transversion rate ratio from the dataset and use this estimate for the parameterization of the model.

```{r sequence-comparison, echo=FALSE}

mt_cyb_df %>%
  mutate("human-mouse" = paste0(human, "-", mouse)) %>%
  group_by(`human-mouse`) %>%
  summarize(n = n()) %>%
  kable(caption = "Nucleotide comparisons") %>%
  kable_styling(full_width = F)


```

