---
title: |
  | Modeling transition/transversion bias
  | of nucleotide substitution over time
subtitle: |
  | Stefan Schmutz
  | January 2019

bibliography: bibliography.bib
csl: nature.csl
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)

library(tidyverse)
library(seqinr)
library(knitr)
library(kableExtra)
```

```{r read-fasta, echo=FALSE}
nt_order <- c("T", "C", "A", "G")

mt_cyb <-
  read.fasta(file = "mt-cyb-human-mouse_cDNAalignment.fasta", set.attributes = FALSE)

mt_cyb_df <-
  bind_cols("mouse" = mt_cyb$MUSMUSCULUSCYTB, "human" = mt_cyb$HOMOSAPIENSCYTB) %>%
  mutate(mouse = fct_relevel(toupper(mouse), nt_order),
         human = fct_relevel(toupper(human), nt_order))
```

## Introduction
DNA, the molecule essential for known life, consists of four building blocks. 
Those blocks (thymine (T), cytosine (C), adenine (A), guanine (G)) are called nucleotides. [@wiki:DNA]  
Since DNA evolves over time, substitutions (change of nucleotides) occur. 
For structural reasons, transitions (substitutions between A and G or T and C), 
are more likely to happen compared to transversions (all other substitutions)
in the majority of the cases [@10.1371/journal.pgen.0030022].

The aim of this work is to describe a Markov model of nucleotide substitution 
over time which consideres nucleotide- and transition/transversion-bias.  
The four nucleotides of DNA represent the four different states of the Markov model.  
Other parameters needed for the model are the steady-state distribution ($\pi$) and 
state transition probabilities. Those parameters will be estimated in the next section 
and are based on a pairwise sequence alignment of human and mouse cytochrome b (MT-CYB) 
which is a mitochondrial gene [@doi:10.1056/NEJM199909303411404].

## Estimations to parameterize the model
There are several different Markov models of nucleotide substitution described in the literature [@doi:10.1093/acprof:oso/9780199602605.001.0001]. The HKY85 model [@Hasegawa1985] was chosen because it fits the problem at hand. The substitution rate matrix $Q$ is defined as follows:  

$$
Q=
\begin{bmatrix}
. & \alpha \pi_{C} & \beta \pi_{A} & \beta \pi_{G}\\
\alpha \pi_{T} & . & \beta \pi_{A} & \beta \pi_{G} \\
\beta \pi_{T} & \beta \pi_{C} & . & \alpha \pi_{G} \\
\beta \pi_{T} & \beta \pi_{C} & \alpha \pi_{A} & .
\end{bmatrix}.
$$

$\alpha$ (transition-rate) and $\beta$ (transversion rate) can be summarised as the transition/transversion rate ratio $\kappa=\alpha/\beta$. The substitution rate matrix can therefore also be described as follows:

$$
Q=
\begin{bmatrix}
. & \kappa \pi_{C} & \pi_{A} & \pi_{G}\\
\kappa \pi_{T} & . & \pi_{A} & \pi_{G} \\
\pi_{T} & \pi_{C} & . & \kappa \pi_{G} \\
\pi_{T} & \pi_{C} & \kappa \pi_{A} & .
\end{bmatrix}.
$$

The diagonal of $Q$ is defined by the requirement, that each row of a substitution rate matrix must sum to 0.

### Nucleotide frequencies

> Estimate nucleotide frequencies from the pairwise alignment of human and mouse cytochrome b gene as given in the file “mt-cyb-human-mouse_cDNAalignment.fasta”. Use these values to parameterize the model.

A way to estimate the nucleotide frequencies is to count the occurences and
divide them by the total length.  
Since we're working with a pairwise alignment without indels, the total length
of both sequences is the same (`r length(mt_cyb$HOMOSAPIENSCYTB)` nt). 
The detailed composition is listed in Table \ref{tab:nucleotide-frequencies}.  


```{r nucleotide-frequencies, echo=FALSE}

nt_frequencies <-
  mt_cyb_df %>%
  pivot_longer(c(mouse, human), names_to = "organism", values_to = "nucleotide") %>%
  group_by(organism, nucleotide) %>%
  summarize(count = n(),
            frequency = round(count / length(mt_cyb$HOMOSAPIENSCYTB), digits = 4))

nt_frequencies %>%
  pivot_wider(-count, names_from = "organism", values_from = "frequency") %>%
  kable(caption = "Nucleotide frequencies") %>%
  kable_styling(full_width = F, latex_options = "hold_position")

avg_frequencies <-
  nt_frequencies %>%
  group_by(nucleotide) %>%
  summarise(avg_frequency = mean(frequency)) %>%
  pull(avg_frequency)

```

We assume that the mean of this distribution is the steady-state distribution $\pi$.  

$$\pi = (\pi_{T}, \pi_{C}, \pi_{A}, \pi_{G}) = (`r round(avg_frequencies, digits = 4)`)$$

### Transition transversion rate ratio

> Propose a simple way of estimating transition transversion rate ratio from the dataset and use this estimate for the parameterization of the model.

The numbers of the 16 possible combinations of the sequence
alignment are shown in Table \ref{tab:nucleotide-comparisons} (from mouse in rows to human in columns).

```{r nucleotide-comparisons, echo=FALSE}

nt_comparisons <-
  mt_cyb_df %>%
  group_by(mouse, human) %>%
  summarize(count = n())

nt_comparisons_matrix <-
  matrix(nt_comparisons$count, nrow = 4, byrow = TRUE, dimnames = list(nt_order, nt_order))

nt_comparisons_matrix %>%
  kable(caption = "Nucleotide comparisons (numbers)") %>%
  kable_styling(full_width = F, latex_options = "hold_position")
```

We get the frequencies (Table \ref{tab:nucleotide-comparisons-frequencies}) when the numbers are divided by the total length (`r length(mt_cyb$HOMOSAPIENSCYTB)` nt).

```{r nucleotide-comparisons-frequencies, echo=FALSE}

round(nt_comparisons_matrix / length(mt_cyb$HOMOSAPIENSCYTB), digits = 4) %>%
  kable(caption = "Nucleotide comparisons (frequencies)") %>%
  kable_styling(full_width = F, latex_options = "hold_position")
```

```{r rate-ratios, echo=FALSE}

transition_difference <-
  nt_comparisons %>%
  filter(mouse == "T" & human == "C" |
         mouse == "C" & human == "T" |
         mouse == "A" & human == "G" |
         mouse == "G" & human == "A") %>%
  pull(count) %>%
  sum() / length(mt_cyb$HOMOSAPIENSCYTB)
  
transversion_difference <-
  nt_comparisons %>%
  filter(mouse == "T" & human == "A" |
         mouse == "A" & human == "T" |
         mouse == "T" & human == "G" |
         mouse == "G" & human == "T" |
         mouse == "C" & human == "A" |
         mouse == "A" & human == "C" |
         mouse == "C" & human == "G" |
         mouse == "G" & human == "C") %>%
  pull(count) %>%
  sum() / length(mt_cyb$HOMOSAPIENSCYTB)

kappa_hat <- 2*log(1-2*transition_difference-transversion_difference)/log(1-2*transversion_difference)-1

```

The fraction of sites with transitional differences ($S = `r round(transition_difference, digits = 4)`$)
and transversional differences ($V = `r round(transversion_difference, digits = 4)`$) can be estimated
by taking the sum of the corresponding fields from Table \ref{tab:nucleotide-comparisons-frequencies}.  
The proportion of different sites ($\hat{p} = S+V = `r round(transition_difference+transversion_difference, digits = 4)`$) 
usually underestimates the amount of substitutions which occured.  
To get a better estimate, we can apply the distance formulae for the K80 model 
which consideres different transitions- transversion-rates [@doi:10.1093/acprof:oso/9780199602605.001.0001].  

The estimate of transition/transversion rate ratio ($\hat{\kappa}$) can now be calculated as follows:  
$$\hat{\kappa} = \frac{2ln(1-2S-V)}{ln(1-2V)}-1 = `r round(kappa_hat, digits = 4)`$$


## (log-)likelihood
### Computing by hand

> By hand, compute the (log-)likelihood for the first 10 alignment positions given that the two sequences are separated by evolutionary distance of 0.01 expected substitutions per site.

To compute the likelihood ($L$) and log-likelihood ($\ell$) we're given that the distance of the two sequences 
from human and mouse is $\hat{d} = 0.01$. This means that the predicted actual 
mutations between the two sequences is 0.01 per position.  
**?how do we use that information?**

The likelihood at a position is defined as the sum of the probabilities of all 
possible ancestral nucleotides. For this pairwise alignment we assume one internal 
node (one common ancestor) which gives us 4 possibilities.  

For this example, likelihood is defined as the product of the likelihoods  
of each of the 10 positions:  
$$L = \prod_{j = 1}^{N} L_{(j)}$$

To avoid underflow (computer issue caused by very small numbers) log-likelihood 
is often used:  
$$\ell = \sum_{j = 1}^{N} ln(L_{(j)})$$

### Computing using a program

> Write a program to compute the likelihood function of the whole alignment for the same genetic distance. Compute the likelihood for the whole alignment for several values of transition-transversion ratio around the value you estimated. Can you find a value that gives a better likelihood?

## Implementing a simulation

> Implement a simulation under your model for two sequences over an arbitrary distance t. Validate the program by simulation and show the results of you validation.

## References
